File: examples/c/custom_draw.c
/*
 * custom_draw.c - Canvas/custom draw API example (C)
 *
 * Demonstrates the immediate-mode canvas API: begin_draw to get a canvas
 * handle, draw shapes and text, then end_draw to flush to the surface.
 * Renders a bar chart similar to the Rust custom_draw example.
 *
 * Build (MSVC Developer Command Prompt):
 *   cl /W4 /I ..\..\crates\winpane-ffi\include custom_draw.c ^
 *      /link /LIBPATH:..\..\target\debug winpane.lib
 *
 * Run:
 *   custom_draw.exe
 */

#include "winpane.h"
#include <stdio.h>
#include <windows.h>

/* Helper: check return code and print last error on failure. */
static int check(int rc, const char *fn) {
    if (rc < 0) {
        /* NOTE: the pointer from winpane_last_error() is invalidated
         * by the next winpane call on the same thread. */
        const char *err = winpane_last_error();
        fprintf(stderr, "%s failed (%d): %s\n", fn, rc, err ? err : "unknown");
    }
    return rc;
}

int main(void) {
    WinpaneContext *ctx = NULL;
    WinpaneSurface *hud = NULL;
    WinpaneCanvas *canvas = NULL;

    /* 1. Create context */
    if (check(winpane_create(&ctx), "winpane_create") < 0)
        return 1;

    /* 2. Create HUD */
    winpane_hud_config_t config = {0};
    config.version = WINPANE_CONFIG_VERSION;
    config.size = sizeof(config);
    config.x = 200;
    config.y = 200;
    config.width = 400;
    config.height = 300;

    if (check(winpane_hud_create(ctx, &config, &hud), "winpane_hud_create") < 0) {
        winpane_destroy(ctx);
        return 1;
    }

    /* 3. Add retained-mode background rect */
    winpane_rect_element_t bg = {0};
    bg.x = 0.0f;
    bg.y = 0.0f;
    bg.width = 400.0f;
    bg.height = 300.0f;
    bg.fill = (winpane_color_t){15, 15, 25, 220};
    bg.corner_radius = 8.0f;
    bg.has_border = 1;
    bg.border_color = (winpane_color_t){60, 60, 100, 180};
    bg.border_width = 1.0f;
    bg.interactive = 0;
    check(winpane_surface_set_rect(hud, "bg", &bg), "set_rect(bg)");

    /* 4. Show surface, brief sleep for window to appear */
    check(winpane_surface_show(hud), "surface_show");
    Sleep(200);

    /* 5. Begin custom draw session */
    if (check(winpane_surface_begin_draw(hud, &canvas), "begin_draw") < 0) {
        winpane_surface_destroy(hud);
        winpane_destroy(ctx);
        return 1;
    }

    /* -- Title -- */
    check(winpane_canvas_draw_text(canvas, 20.0f, 15.0f,
          "Weekly Activity", 18.0f,
          (winpane_color_t){255, 255, 255, 255}), "draw_text(title)");

    /* -- Horizontal baseline -- */
    check(winpane_canvas_draw_line(canvas,
          40.0f, 240.0f, 370.0f, 240.0f,
          (winpane_color_t){80, 80, 120, 200}, 1.0f), "draw_line(baseline)");

    /* -- Bar chart -- */
    {
        const float bar_values[4] = {0.7f, 0.45f, 0.9f, 0.3f};
        const char *bar_labels[4] = {"Mon", "Tue", "Wed", "Thu"};
        const winpane_color_t bar_colors[4] = {
            {80, 160, 255, 255},
            {100, 220, 160, 255},
            {255, 180, 80, 255},
            {255, 100, 120, 255},
        };
        const float bar_width = 60.0f;
        const float bar_max_height = 170.0f;
        const float start_x = 55.0f;
        const float spacing = 80.0f;
        int i;

        for (i = 0; i < 4; i++) {
            float x = start_x + (float)i * spacing;
            float bar_height = bar_values[i] * bar_max_height;
            float y = 240.0f - bar_height;

            /* Filled rounded rect for the bar */
            check(winpane_canvas_fill_rounded_rect(canvas,
                  x, y, bar_width, bar_height, 4.0f,
                  bar_colors[i]), "fill_rounded_rect(bar)");

            /* Label below bar */
            check(winpane_canvas_draw_text(canvas,
                  x + 15.0f, 248.0f,
                  bar_labels[i], 12.0f,
                  (winpane_color_t){160, 160, 180, 255}), "draw_text(label)");

            /* Value above bar */
            {
                char value_str[16];
                snprintf(value_str, sizeof(value_str), "%d%%",
                         (int)(bar_values[i] * 100.0f));
                check(winpane_canvas_draw_text(canvas,
                      x + 12.0f, y - 20.0f,
                      value_str, 11.0f,
                      bar_colors[i]), "draw_text(value)");
            }
        }
    }

    /* -- Decorative elements -- */

    /* Stroke ellipse in top-right corner */
    check(winpane_canvas_stroke_ellipse(canvas,
          370.0f, 30.0f, 12.0f, 12.0f,
          (winpane_color_t){100, 180, 255, 120}, 1.5f), "stroke_ellipse");

    /* Stroke rect border around chart area */
    check(winpane_canvas_stroke_rect(canvas,
          30.0f, 45.0f, 350.0f, 220.0f,
          (winpane_color_t){40, 40, 70, 100}, 1.0f), "stroke_rect(border)");

    /* 6. End draw - flush ops to surface */
    /* NOTE: canvas handle is invalid after this call. */
    check(winpane_surface_end_draw(hud), "end_draw");
    canvas = NULL;

    printf("winpane custom_draw: bar chart overlay at (200, 200).\n");
    printf("Press Ctrl+C to exit.\n");

    /* 7. Sleep loop to keep the overlay visible */
    for (;;) {
        Sleep(1000);
    }

    /* 8. Cleanup (unreachable in this example, but shown for correctness) */
    winpane_surface_destroy(hud);
    winpane_destroy(ctx);
    return 0;
}

File: examples/rust/custom_draw.rs
//! Demo: custom draw escape hatch
//!
//! Creates a HUD and uses DrawOp to render custom content beyond
//! the retained-mode scene graph: filled rects, text, lines, ellipses.
//!
//! Run on Windows: cargo run -p winpane --example custom_draw

// ── winpane design tokens ──────────────────────────────────────
// Surface base:   rgb(18, 18, 22)  Glass: a=228  Solid: a=255  Muted: a=242
// Elevated:       rgb(28, 28, 33)  Interactive:  rgba(38, 38, 44, 255)
// Border:         rgba(255,255,255, 18)     Hover:       rgba(48, 48, 56, 255)
// Text primary:   rgba(232, 232, 237, 255)  Secondary:   rgba(148, 148, 160, 255)
// Accent:         rgba(82, 139, 255, 255)   Accent hover:rgba(110, 160, 255, 255)
// Success:        rgba(52, 211, 153, 255)   Warning:     rgba(251, 191, 36, 255)
// Danger:         rgba(239, 68, 68, 255)    Radius: 10/6 px
// ────────────────────────────────────────────────────────────────

use winpane::{Color, Context, DrawOp, HudConfig, Placement, RectElement};

#[allow(clippy::print_stdout)]
fn main() -> Result<(), winpane::Error> {
    let ctx = Context::new()?;

    let hud = ctx.create_hud(HudConfig {
        placement: Placement::Position { x: 200, y: 200 },
        width: 400,
        height: 300,
    })?;

    // Add a retained-mode background
    hud.set_rect(
        "bg",
        RectElement {
            x: 0.0,
            y: 0.0,
            width: 400.0,
            height: 300.0,
            fill: Color::rgba(18, 18, 22, 228),
            corner_radius: 10.0,
            border_color: Some(Color::rgba(255, 255, 255, 18)),
            border_width: 1.0,
            interactive: false,
        },
    );

    hud.show();

    // Give the window time to appear
    std::thread::sleep(std::time::Duration::from_millis(200));

    // Custom draw: bar chart with labels
    let bar_colors = [
        Color::rgba(82, 139, 255, 255),
        Color::rgba(52, 211, 153, 255),
        Color::rgba(251, 191, 36, 255),
        Color::rgba(239, 68, 68, 255),
    ];
    let bar_values: [f32; 4] = [0.7, 0.45, 0.9, 0.3];
    let bar_labels = ["Mon", "Tue", "Wed", "Thu"];

    let mut ops = Vec::new();

    // Title
    ops.push(DrawOp::DrawText {
        x: 20.0,
        y: 15.0,
        text: "Weekly Activity".into(),
        font_size: 16.0,
        color: Color::rgba(232, 232, 237, 255),
    });

    // Horizontal baseline
    ops.push(DrawOp::DrawLine {
        x1: 40.0,
        y1: 240.0,
        x2: 370.0,
        y2: 240.0,
        color: Color::rgba(255, 255, 255, 18),
        stroke_width: 1.0,
    });

    // Bars
    let bar_width = 60.0;
    let bar_max_height = 170.0;
    let start_x = 55.0;
    let spacing = 80.0;

    for (i, (&value, &color)) in bar_values.iter().zip(bar_colors.iter()).enumerate() {
        let x = start_x + i as f32 * spacing;
        let bar_height = value * bar_max_height;
        let y = 240.0 - bar_height;

        ops.push(DrawOp::FillRoundedRect {
            x,
            y,
            width: bar_width,
            height: bar_height,
            radius: 4.0,
            color,
        });

        // Label below bar
        ops.push(DrawOp::DrawText {
            x: x + 15.0,
            y: 248.0,
            text: bar_labels[i].into(),
            font_size: 13.0,
            color: Color::rgba(148, 148, 160, 255),
        });

        // Value above bar
        ops.push(DrawOp::DrawText {
            x: x + 12.0,
            y: y - 20.0,
            text: format!("{}%", (value * 100.0) as u32),
            font_size: 11.0,
            color,
        });
    }

    // Decorative ellipse
    ops.push(DrawOp::StrokeEllipse {
        cx: 370.0,
        cy: 30.0,
        rx: 12.0,
        ry: 12.0,
        color: Color::rgba(82, 139, 255, 120),
        stroke_width: 1.5,
    });

    hud.custom_draw(ops);

    println!("winpane custom_draw: overlay with bar chart at (200, 200).");
    println!("Press Ctrl+C to exit.");

    loop {
        std::thread::sleep(std::time::Duration::from_secs(1));
    }
}
